src/ahip/connection.py

```diff
@@ -134,63 +115,83 @@ def _make_body_iterable(body):
     is deliberate: users must make choices about the encoding of the data they
     use.
     """
-    if body is None:
-        return []
-    elif isinstance(body, bytes):
-        return [body]
-    elif hasattr(body, "read"):
-        return _read_readable(body)
-    elif isinstance(body, collections.Iterable) and not isinstance(body, six.text_type):
-        return body
-    else:
-        raise InvalidBodyError("Unacceptable body type: %s" % type(body))
+
+    async def generator():
+        if body is None:
+            return
+        elif isinstance(body, bytes):
+            yield body
+        elif hasattr(body, "read"):
+            async for chunk in _read_readable(body):
+                yield chunk
+        elif isinstance(body, collections.Iterable) and not isinstance(
+            body, six.text_type
+        ):
+            for chunk in body:
+                yield chunk
+        else:
+            raise InvalidBodyError("Unacceptable body type: %s" % type(body))
+
+    return generator().__aiter__()
 
 
-# XX this should return an async iterator
 def _request_bytes_iterable(request, state_machine):
     """
     An iterable that serialises a set of bytes for the body.
     """
 
     def all_pieces_iter():
```

Comment:
I'm not super happy with this function, there's probably a simpler way to do this rather than a direct translation.

Reply:
Possibly. Maybe @njsmith know how to improve this? In any case, this isn't blocking, so I'll merge this pull request.